### PR TITLE
add support for RC in AP_Periph

### PR DESCRIPTION
**Pull request by peterbarker**
This currently only allows for instantiation of RC in periph.

We still need a DroneCAN packet format, and an implementation of that both for AP_Periph (sending side) and `AP_RCProtocol_DroneCAN` (receiving side).

- [x] SBUS test (OrangeRX, throttle-based failsafe only)
- [x] FPort test (FS-A8S)
- [x] DSM test
- [x] failover test
- [x] recovery test
- [x] no-pulses vs SBUS failsafe test

### CODE REVIEW

File: Tools/AP_Periph/AP_Periph.h [hunk 1]
```diff
@@ -30,6 +30,7 @@
 #if HAL_WITH_ESC_TELEM
 #include <AP_ESC_Telem/AP_ESC_Telem.h>
 #endif
+#include <AP_RCProtocol/AP_RCProtocol.h>
```
Comment (peterbarker) [inline Tools/AP_Periph/AP_Periph.h:33]: Not needed in header?

**Comment by peterbarker** [comment]
Conversation from Discord:

```
11:55 AM]Enki: Could be something like this. I think this would be very flexible, while also simple to configure, and simple to read and write packets. You could use the GetSet, but that is a bit messy to code for, as I discovered; would prefer sending and receiving a custom config packet that has these fields.

/// Configuration for this receiver node. Can be read and set over CAN, or USB.
/// Not currently in DroneCAN DSDL; addition pending. 
pub struct Config {
    /// Includes common CAN config, like bitrate, and if FD mode (64-byte packets) is enabled.
    pub common: ConfigCommon,
    /// Divide the ELRS update rate transmitted over the radio link by this. So, 1 means transmit every
    /// RC packet received. 2 means transmit every other update, etc.
    pub broadcast_ratio: u8,
    /// Up to 3 grades of precision, labeled `a`, `b`, and `c`. These `precision` settings
    /// are in bit-depth, and can be up to 64-bits. Broadcast order is all a channels, all b channels,
    /// all c channels, with no padding between. The indices are deteremined entirely by precision,
    /// and number of channels.
    /// 
    /// An example: Use 4 a channels as high-precision for control data. Use b channels as low-precision
    /// for switch and button positions, eg for arming or setting flight mode.
    pub ch_type_a_precision: u8,
    pub ch_type_b_precision: u8,
    pub ch_type_c_precision: u8,
    /// Number of channels per precision level.
    pub num_a_channels: u8,
    pub num_b_channels: u8,
    pub num_c_channels: u8,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            common: Default::default(),
            broadcast_ratio: 1,
            ch_type_a_precision: 12,
            ch_type_b_precision: 4,
            ch_type_c_precision: 1,
            num_a_channels: 4,
            num_b_channels: 10,
            num_c_channels: 0,
        }
    }
}
[11:55 AM]Enki: Maybe something like that?
[12:00 PM]Peter Barker: Huh.  Much more complicate than I'd imagined   I was thinking something like 8 11-bit values + a 7-bit RSSI + 2 bits of "bank" - and 7 bits for random other things, that sort of thing.
[12:01 PM]Enki: Haven't even considered the link stats yet...
[12:01 PM]Peter Barker: In your scenario, how do you keep the RC receiver configuration in-step with the user's RC_Channels (and, sadly, RCMap) configuration? (edited)
[12:02 PM]Peter Barker: Protip: "they have to configure it" is probably not an option for most scenarios 
[12:02 PM]Enki: I think the sensible-defaults would help there
[12:02 PM]Enki: Don't have a good answer on syncing with the FC firmware
[12:02 PM]Enki: The FC could negotiate over CAN, but that adds complication
[12:03 PM]Enki: Maybe the answer is lean heavily on the defaults, but have the node fundamentally capable of flexibility
[12:03 PM]Enki: Ie maybe that default config above (it's somewhat arbitrary as-is) would be used in 99% of cases
[12:04 PM]Peter Barker: What does a data packet look like in such a scheme?  Essentially "here's a whole bunch of bits, you know what's going on"?
[12:04 PM]Enki: Config packet, or channel data?
[12:05 PM]Peter Barker: channel data
[12:06 PM]Enki: Using the default I posted:
12 bits of channel 1
12 bits of channel 2
12 bits of channel 3
12 bits of channel 4
4 bits of channel 5
4 bits of channel 6
(8 more 4-bit channels)
[12:06 PM]Enki: No padding
[12:06 PM]Enki: (Not byte-aligned unfortunately)
[12:07 PM]Peter Barker: Right, but it's up to the user of the packet to know what's in it, so the only thing the DroneCAN packet can be is "lump of bits containing data packed according to node configuration".
[12:07 PM]Enki: Ways the user could know
- Datasheet that shows the default
- Reading the config over CAN
[12:07 PM]Enki: - Plugging the device into a PC and reading using the included PC software
[12:08 PM]Peter Barker: Yep.  Just making sure I know what's going on.  I think what I'm saying is right.
[12:08 PM]Enki: In case it's not clear, I'm making this up as I go!
[12:09 PM]Enki: Would love to get input from diff people with diff use cases
[12:09 PM]Enki: Will probably continue prototyping using CRSF initially
@Enki
Will probably continue prototyping using CRSF initially
[12:13 PM]Peter Barker: Yep.  I've been trying not to think about the bi-directional telemetry stuff when thinking about this - do you see a compelling reason to couple telemetry (including mavlink-passthrough commands) to receiving of RC signals?  I'm thinking separate data packets on the CANBus are the right thing to do there.  Perhaps interpret the CRSF telemetry (or SPort) on the periph node and translate into a common packet format on the CANBus.
[12:13 PM]Enki: I concur re decoupling
[12:13 PM]Enki: 2 unrelated concepts IMO
[12:13 PM]Enki: other than they are both RF data between ground station and aircraft
[12:15 PM]Enki: *Eh I guess it depends on what messages specifically are going to/from
```

**Comment by peterbarker** [comment]
```
1:26 PM]Enki: My latest thought is still this for a config packet, with 3 classes of channels, where precision is user-defined:
/// Configuration for this receiver node. Can be read and set over CAN, or USB.
/// Not currently in DroneCAN DSDL; addition pending.
pub struct Config {
    /// Includes common CAN config, like bitrate, and if FD mode (64-byte packets) is enabled.
    pub common: ConfigCommon,
    /// Divide the ELRS update rate transmitted over the radio link by this. So, 1 means transmit every
    /// RC packet received. 2 means transmit every other update, etc.
    pub broadcast_ratio: u8,
    /// Up to 3 grades of precision, labeled `a`, `b`, and `c`. These `precision` settings
    /// are in bit-depth, and can be up to 64-bits. Broadcast order is all a channels, all b channels,
    /// all c channels, with no padding between. The indices are deteremined entirely by precision,
    /// and number of channels.
    ///
    /// An example: Use 4 a channels as high-precision for control data. Use b channels as low-precision
    /// for switch and button positions, eg for arming or setting flight mode.
    pub ch_type_a_precision: u8,
    pub ch_type_b_precision: u8,
    pub ch_type_c_precision: u8,
    /// Number of channels per precision level.
    pub num_a_channels: u8,
    pub num_b_channels: u8,
    pub num_c_channels: u8,
}
[1:26 PM]Enki: But, that may be too flexible
[1:36 PM]Enki: Another approach is fixed-size but generous (at the expensve of packet-size), like u16 high-precision channels
[1:42 PM]Peter Barker: Why the extra precision?  11 bits ought to be enough for anybody   I assume you think those extra 5 bits are worth-while?  That an operator can move their sticks in sub-0.1% increments?  And notice the difference? (edited)
NEW
[2:03 PM]Enki: I'm trying to think of the broader use case
[2:04 PM]Enki: The unknown unknowns
[2:04 PM]Enki: I see your point though. Need to keep this grounded in practical use cases
[2:05 PM]Enki: Maybe 12
[2:06 PM]Enki: Given your point about hand-adjusted channel data not needing more than 11 or so bits, what are your thoughts on handling switch, button etc channels?
[2:13 PM]Enki: Would you do something like:

struct Config {
    /// Number of 12-bit channels, ie for control-channel-data
    num_high_precision_channels: u8,
    num_5_bit_channels: u8,
    num_4_bit_channels: u8,
    num_3_bit_channels: u8,
    num_2_bit_channels: u8,
    num_1_bit_channels: u8,
}


Where the packet is arranged, as before, with no padding, in the order prescribed by a config as above
[2:14 PM]Enki: Something like that might strike a nice balance
```

**Comment by David-OConnor** [comment]
Update from more convo on Discord: Peter has perhaps talked me off the ledge of arbitrary channel precision, with the justification that control channels, for practical purposes, may not need more than 11 (or 12?) bits. I'm unable to come up with a counter-example.

My latest best-shot at what a config looks like may be something like this:

```rust
struct Config {
    /// Divide the update rate transmitted over the radio link by this. So, 1 means transmit every
    /// RC packet received. 2 means transmit every other update, etc. 0 means don't transmit.
    broadcast_ratio: u8,
    /// Number of 12-bit channels, ie for control-channel-data
    num_high_precision_channels: u8,
    /// Number of 5-bit channels, eg for selecting from a number of options
    num_5_bit_channels: u8,
    num_4_bit_channels: u8,
    /// Multi-position switches. For example, for setting flight mode
    num_3_bit_channels: u8,
    /// Eg 3-position switches
    num_2_bit_channels: u8,
    /// Eg 2-position switches
    num_1_bit_channels: u8,
}    
```

Data would be transmitted in the order listed, with no padding. So, if configured for `4` high-precision (say, 11-bit) channels and `2` 1-bit channels, the first 44 bits would be control channel data, and the last 2 bits would be 1-bit channels.

**Comment by peterbarker** [comment]
This is an example of a minimalist RC input packet.  Notionally we could send 5 11 bit values in a single packet with this same format but nobody uses only 5 channels.

```
#
# support for RC input
#

# quality metric, -1-100, -1 meaning unknown, 0 meaning no RC input
#  being received by this sensor

int7_t quality

# bank of input.  Find channel by multiplying this number by 6 and adding input number:
uint2 bank 

# input1; if bank is zero this is channel 1.  Find value by adding 500
to this value.
uint11 input1

# more RC inputs:
uint11 input2
uint11 input3
uint11 input4
uint11 input5
uint11 input6
uint11 input7
uint11 input8
uint11 input9
uint11 input10


# 7 + 2 + (10*11) == 119 bits == 14bytes+7 bits == 2 packets  (9 bits spare)

# could go with 9 12-bits or 10 bits of precision.
```

**Comment by tridge** [comment]
how about this: ```
uint8 STATUS_RSSI_VALID=1 # RSSI field is valid
uint8 STATUS_FAILSAFE=2 # receiver has lost contact with transmitter
uint8 status

uint8 rssi
uint12[<=24] rcin
```

**Comment by David-OConnor** [comment]
I think RSSI would be a better fit for a separate link stats packet type. It would include info like RSSI, % of recent packets received, signal strength, xmitter power, active antenna etc.

**Comment by peterbarker** [comment]
> 

So for 10 channels coming in:

>>> (8 + 8 + 10*12)/8.0
17.0
>>> 

... just a little too much?

**Comment by tridge** [comment]
> So for 10 channels coming in:

it is just as likely to be 12 or 16 channels

**Comment by tridge** [comment]
maybe uint16 status, so we can support things like "STATUS_SUPPORTS_FRSKY" etc

**Comment by cbf123** [comment]
This is for an on-the-wire protocol, yes?  Unless bandwidth is a major factor, why not just keep it simple and have an arbitrary number of 12-bit channels plus a header for random stuff?

And I think we'd want to support diversity receivers, so two separate rssi values and an "active antenna" bit might make sense.

**Comment by David-OConnor** [comment]
After some thinking, an arbitrary num of 12-bit channels would be fine. Would be a bit heavy if some of the channels are for 2-bit switches, but maybe that's OK, and better from a simplicity perspective. Or agree on a packing protocol between FC firmware and Rx nodes as a backup.

**Comment by peterbarker** [comment]
> This is for an on-the-wire protocol, yes? Unless bandwidth is a major factor, why not just keep it simple and have an arbitrary number of 12-bit channels plus a header for random stuff?
> 
> And I think we'd want to support diversity receivers, so two separate rssi values and an "active antenna" bit might make sense.

This one does have a bunch of 12-bit values.  Interesting point on the dual-rssi values - which RC protocols bare that information?

**Comment by MallikarjunSE** [comment]
Doesn't periph require sbus inverter? Very few periphs with H7 breakouts.

**Comment by cbf123** [comment]
>  Interesting point on the dual-rssi values - which RC protocols bare that information?

Pretty sure CRSF (used by Crossfire and ELRS and maybe others) does.

ELRS recently merged https://github.com/ExpressLRS/ExpressLRS/pull/2143

**Comment by peterbarker** [comment]
I've flown this on a quadcopter today; MatekL431 node with a Turnigy iBus receiver hanging off it.

**Comment by peterbarker** [comment]
> Doesn't periph require sbus inverter? Very few periphs with H7 breakouts.

We're leaning on AP_RCProtocol to handle serial negotiations.  Inversion is done in `AP_RCProtocol::SerialConfig::apply_to_uart`. My reading is that F3, G4, L4 and L4+ will all invert, so there are options :-)

**Comment by David-OConnor** [comment]
Hi! My WIP node uses a  G4 MCU for the node, but I'm not following this. What is an sbus inverter, in this context? Thank you!

File: libraries/AP_RCProtocol/AP_RCProtocol.cpp [hunk 2]
```diff
@@ -367,6 +368,24 @@ bool AP_RCProtocol::new_input()
             backend[i]->update();
         }
     }
+
+#if AP_RCPROTOCOL_DRONECAN_ENABLED
+    uint32_t now = AP_HAL::millis();
+    if (should_search(now)) {
+        if (backend[AP_RCProtocol::DRONECAN]->new_input()) {
```
Comment (tridge) [inline libraries/AP_RCProtocol/AP_RCProtocol.cpp:375]: should we check if that backend is allocated here? just a nullptr check

File: libraries/AP_RCProtocol/SoftSerial.cpp [hunk 3]
```diff
@@ -85,9 +85,17 @@ bool SoftSerial::process_pulse(uint32_t width_high, uint32_t width_low, uint8_t
         }
         if (config == SERIAL_CONFIG_8E2I) {
             // check parity
+#ifdef HAL_BUILD_AP_PERIPH
+            const uint8_t popcount = __builtin_popcount((state.byte>>1)&0xFF);
```
Comment (tridge) [inline libraries/AP_RCProtocol/SoftSerial.cpp:89]: we need to know why __builtin_parity() fails on periph

**Comment by cbf123** [comment]
> Hi! My WIP node uses a G4 MCU for the node, but I'm not following this. What is an sbus inverter, in this context? What does [STM32?]H7 have to do with it? Thank you!

SBUS is a "normally inverted" protocol, so using it with a hardware UART requires either that the UART can itself invert the signal (which F3, F7, H7, etc. processors can do) or that a hardware SBUS inverter be used to invert the signal before it reaches the UART (on an F4 processor, for example).

**Comment by David-OConnor** [comment]
Thank you. My understanding is that is an implementation detail of the node, so shouldn't affect what AP sees from a CAN device. I think I am likely missing some context.

**Comment by peterbarker** [comment]
> Hi! My WIP node uses a G4 MCU for the node, but I'm not following this. What is an sbus inverter, in this context? What does [STM32?]H7 have to do with it? Thank you!

SBUS is an inverted protocol - up is down and down is up in terms of the on-wire bits.  There are "non-inverted" variants of that protocol.  Some autopilots have a hardware inverter - e.g. PixRacer.  Some stm32 processors allow you to toggle it on the MCU - I listed some.

I don't think you need to worry about it for your hardware - ArduPilot should work it out by trying both inverted and then non-inverted protocols.

**Comment by David-OConnor** [comment]
> > Hi! My WIP node uses a G4 MCU for the node, but I'm not following this. What is an sbus inverter, in this context? What does [STM32?]H7 have to do with it? Thank you!
> 
> SBUS is an inverted protocol - up is down and down is up in terms of the on-wire bits. There are "non-inverted" variants of that protocol. Some autopilots have a hardware inverter - e.g. PixRacer. Some stm32 processors allow you to toggle it on the MCU - I listed some.
> 
> I don't think you need to worry about it for your hardware - ArduPilot should work it out by trying both inverted and then non-inverted protocols.

My understanding is the reason I don't need to worry about it is because this AP talks to the Rx over CAN (not serial, inverted or not), no?

**Comment by peterbarker** [comment]
> > I don't think you need to worry about it for your hardware - ArduPilot should work it out by trying both inverted and then non-inverted protocols.
> 
> My understanding is the reason we don't need to worry about it is because AP talks to the Rx over CAN (not serial, inverted or not), no?

... the AP_Periph node is speaking serial to the receiver, however, and *it* needs to know how to invert.  Which it does ;-)

**Comment by MallikarjunSE** [comment]
> Inversion is done in AP_RCProtocol::SerialConfig::apply_to_uart. My reading is that F3, G4, L4 and L4+ will all invert, so there are options :-)

Thanks Peter. I was thinking it was available only on recent ones like H7, G4, L5 series.

**Comment by David-OConnor** [comment]
> > > I don't think you need to worry about it for your hardware - ArduPilot should work it out by trying both inverted and then non-inverted protocols.
> > 
> > 
> > My understanding is the reason we don't need to worry about it is because AP talks to the Rx over CAN (not serial, inverted or not), no?
> 
> ... the AP_Periph node is speaking serial to the receiver, however, and _it_ needs to know how to invert. Which it does ;-)

Thank you for clarifying; I misunderstood the role of this PR; it's for *using AP_PERIPH itself to develop RC node firmware*; not just AP interepreting CAN RC input.

File: libraries/AP_Math/crc.cpp [hunk 4]
```diff
@@ -539,3 +541,30 @@ uint64_t crc_crc64(const uint32_t *data, uint16_t num_words)
...
+
+// return the parity of byte - "1" if there is an odd number of bits
+// set, "0" if there is an even number of bits set note that
+// __builtin_parity causes hardfaults on Pixracer-periph - and is
+// slower on 1 byte than this:
+uint8_t parity(uint8_t byte)
+{
+    uint8_t p = 0;
+
+    p ^= byte & 0x1;
```
Comment (tridge) [inline libraries/AP_Math/crc.cpp:553]: why not p ^= byte & 0x2;

**Review by tridge — CHANGES_REQUESTED** [review]
needs double shuffle on the DSDL submodule

File: libraries/AP_Math/crc.cpp (hunk 4 above)
Comment (peterbarker) [reply libraries/AP_Math/crc.cpp:553]: Tried that's it's bigger and slower as you have to turn the result into a 1 or 0 afterwards

File: Tools/AP_Periph/can.cpp [hunk 5]
```diff
@@ -2702,6 +2702,41 @@ void AP_Periph_FW::can_efi_update(void)
...
 
+#if AP_RCPROTOCOL_ENABLED
+/*
+  send an RCInput CAN message
+ */
+void AP_Periph_FW::can_send_RCInput(uint8_t rssi, uint16_t *values, uint8_t nvalues, bool in_failsafe, bool rssi_valid)
+{
+    uint16_t status = 0;
+    if (rssi_valid) {
+        status |= 1;
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply Tools/AP_Periph/can.cpp:2713]: Fixed

File: Tools/AP_Periph/can.cpp [hunk 6]
```diff
@@ -2702,6 +2702,41 @@ void AP_Periph_FW::can_efi_update(void)
...
+    }
+    if (in_failsafe) {
+        status |= 2;
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply Tools/AP_Periph/can.cpp:2716]: Fixed

**Comment by peterbarker** [comment]
image

**Comment by David-OConnor** [comment]
Hi! It appears after an offline discussion, this PR does 2 things:

1: Adds support for DroneCAN control receivers in Ardupilot
2: Allows AP_Periph to be used to create control receiver node firmware.

In the checklist at the top, are any of the blockers for 1? Would it be feasible to split these 2 so a blocker on one doesn't impact the other? Thank you!

**Comment by peterbarker** [comment]
I've now tested this PR to ensure failover happens between RC input coming from the IO processor and via an AP_Periph node.  I simulated no-pulses failsafe by simply unplugging the receiver and moving it between the peripheral node and the RC input on a Qiotek Zealot H743.

**Comment by peterbarker** [comment]
I've now tested this with anFS-X8B receiver which sets the failsafe bit if it loses connection with the transmitter.

Failsafe and recovery work as expected when connected to the Periph node.

**Comment by peterbarker** [comment]
I've tested DroneCAN input with DSM input via the Spektrum port on a CubeOrangePlus.

the DSM input takes precedence over the DroneCAN input, and if either goes away the other is happy to supply input, and if both go away RC failsafe occurs.

File: Tools/AP_Periph/Parameters.cpp [hunk 7]
```diff
@@ -551,6 +555,35 @@ const AP_Param::Info AP_Periph_FW::var_info[] = {
...
+    // RC_PROTOCOLS copied from RC_Channel/RC_Channels_Varinfo.h
+    // @Param: RC_PROTOCOLS
+    // @DisplayName: RC protocols enabled
+    // @Description: Bitmask of enabled RC protocols. Allows narrowing the protocol detection to only specific types of RC receivers which can avoid issues with incorrect detection. Set to 1 to enable all protocols.
+    // @User: Advanced
+    // @Bitmask: 0:All,1:PPM,2:IBUS,3:SBUS,4:SBUS_NI,5:DSM,6:SUMD,7:SRXL,8:SRXL2,9:CRSF,10:ST24,11:FPORT,12:FPORT2,13:FastSBUS
+    GSCALAR(rcin_protocols, "RC_PROTOCOLS", 1),
+
+    // RC_PROTOCOLS copied from RC_Channel/RC_Channels_Varinfo.h
+    // @Param: RC_RATE
```
Comment (IamPete1) [inline Tools/AP_Periph/Parameters.cpp:568]:
```suggestion
    // @Param: RC_MSGRATE
```

File: Tools/AP_Periph/rc_in.cpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+void AP_Periph_FW::rcin_update()
+{
+    if (!rcin_initialised) {
+        return;
+    }
+
+    auto &rc = AP::RC();
+    if (!rc.new_input()) {
+        return;
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply Tools/AP_Periph/rc_in.cpp:59]: The latter.  We've got nothing to report so I figured don't send anything.  This allows you to convey a little more information with the exitsing flags (i.e. the difference between no pulses and an SBUS failsafe bit being set.

File: Tools/AP_Periph/Parameters.cpp (hunk 7 above)
Comment (peterbarker) [reply Tools/AP_Periph/Parameters.cpp]: > 

Fixed, thanks.

File: Tools/AP_Periph/AP_Periph.cpp [hunk 9]
```diff
@@ -159,6 +159,10 @@ void AP_Periph_FW::init()
     battery.lib.init();
 #endif
 
+#if AP_RCPROTOCOL_ENABLED
```
Comment (tridge) [inline Tools/AP_Periph/AP_Periph.cpp:162]: please use HAL_PERIPH_ENABLE_RCIN

File: Tools/AP_Periph/can.cpp [hunk 10]
```diff
@@ -2706,6 +2706,41 @@ void AP_Periph_FW::can_efi_update(void)
 }
 #endif // HAL_PERIPH_ENABLE_EFI
 
+#if AP_RCPROTOCOL_ENABLED
+/*
```
Comment (tridge) [inline Tools/AP_Periph/can.cpp:2710]: move to rc_in.cpp would be good

**Review by tridge — CHANGES_REQUESTED** [review]

File: Tools/AP_Periph/AP_Periph.cpp (hunk 9 above)
Comment (peterbarker) [reply Tools/AP_Periph/AP_Periph.cpp:162]: Done

**Comment by David-OConnor** [comment]
Is there any testing I can do to help?

**Comment by David-OConnor** [comment]
Not working on my end; no signs of life. LMK what I can do to help.

**Review by tridge — APPROVED** [review]

**Review by tridge — APPROVED** [review]

**Pull request merged**